components/function-controller/internal/gitops/operator.go

```diff
@@ -0,0 +1,133 @@
+package gitops
+
+import (
+	"fmt"
+
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/pkg/errors"
+)
+
+const (
+	usernameKey = "username"
+	passwordKey = "password"
+)
+
+//go:generate mockery -name=GitInterface -output=automock -outpkg=automock -case=underscore
+type GitInterface interface {
+	Clone(s storage.Storer, worktree billy.Filesystem, o *git.CloneOptions) (*git.Repository, error)
+	PlainClone(path string, isBare bool, options *git.CloneOptions) (*git.Repository, error)
+}
+
+type Config struct {
+	RepoUrl string
+	Branch  string
+	BaseDir string
+	Secret  map[string]string
+}
+
+type Operator struct {
+	gitInterface GitInterface
+}
+
+func NewOperator() *Operator {
+	return &Operator{gitInterface: NewGit()}
+}
+
+func (g *Operator) GetLastCommit(config Config) (commitHash string, err error) {
+	auth, err := convertToBasicAuth(config.Secret)
+	if err != nil {
+		return commitHash, errors.Wrap(err, "while parsing auth fields")
+	}
+
+	repo, err := g.gitInterface.Clone(memory.NewStorage(), nil, &git.CloneOptions{
```

Comment:
Did you test it with a large repository with lots of commits? If it clones the whole repository to the memory, then it is not a good idea 😞